### PR TITLE
fix(semantic): model containers as first-class tree nodes for 3-way merge (P0)

### LINKED ISSUE #484: P0: semantic merge silently corrupts subdirectory-file merges (postamble duplication + rebase false-conflict)
**Severity: P0 / critical — demo-blocker.** The function-level / `thread refresh` 3-way merge **silently** corrupts and de-attributes any file under a subdirectory (`src/…`), i.e. essentially all real Rust code. The merge prints `merge type: fast-forward`, emits **0 conflict markers**, and `status` reads clean — yet the result is structurally corrupt and blame collapses onto one merge state. Silent wrong-merges are the highest-risk class: nothing signals the corruption. Verified reproducing on `main` @ `5332f61d`.

## Two composing defects (verified by byte-level probe)

**1. The semantic merge driver is unconditionally buggy on this input shape — path makes no difference.**
`crates/semantic/src/merge_driver/reconstruct.rs:201-235` (`reconstruct_merged_file`): the trailing postamble range (`// MARK`) is pulled in BOTH as the *bridging* preceding-segment for an added item the side lacks (`side_range_for_emit`, reconstruct.rs:286+) AND again as the postamble (223-235). The `emitted: [BTreeSet<usize>; N]` dedup (line 188) does not collapse across the two added items, so the postamble is emitted twice and the second added item lands after it. A direct call to `semantic_three_way_merge(base="a,MARK", ours="a,c,MARK", theirs="a,b,MARK")` returns the corrupt output for `lib.rs`, `src/lib.rs`, and `src/foo/lib.rs` alike.

**2. The subdirectory discriminator lives in the refresh/rebase path, not the driver.**
`refresh_thread` (`crates/cli/src/cli/commands/thread_cmd.rs:432-503`) refreshes via a **rebase** first and only falls back to the semantic 3-way merge **when the rebase flags a conflict** (`try_three_way_merge_refresh` → `apply_merged_tree_external`, thread_cmd.rs:1053). The rebase engine succeeds cleanly for a top-level file (correct output, corrupt semantic preflight discarded) but **spuriously conflicts on a file under a subdirectory**, routing it into the buggy semantic fallback whose corrupt blob is materialized. Blame collapses because the fallback re-snapshots the whole file as one new merge state.

So: subdir discriminator = "rebase falsely conflicts on nested paths"; corruption = "semantic reconstruct duplicates the postamble".

## Repros (only the file path differs; verified)
```bash
trial() {  # $1 = path of the merged file
  d=$(mktemp -d); cd "$d"; git init -qb main; git config user.name a; git config user.email a@a
  mkdir -p "$(dirname "./$1")"
  printf 'pub fn a() {}\n\n// MARK\n' > "$1"
  git add -A; git commit -qm base >/dev/null; heddle adopt --ref main >/dev/null
  heddle thread create x >/dev/null; heddle thread create y >/dev/null
  heddle thread switch x >/dev/null; printf 'pub fn a() {}\n\npub fn b() {}\n\n// MARK\n' > "$1"; heddle commit -m b >/dev/null
  heddle thread switch y >/dev/null; printf 'pub fn a() {}\n\npub fn c() {}\n\n// MARK\n' > "$1"; heddle commit -m c >/dev/null
  heddle thread switch main >/dev/null; heddle merge x -m mx >/dev/null
  heddle thread switch y >/dev/null; heddle thread refresh y >/dev/null
  heddle thread switch main >/dev/null; heddle merge y -m my >/dev/null
  echo "[$1] MARK_lines=$(grep -c MARK "$1") (want 1)  blame_states=$(heddle blame "$1" | grep -oE '^hd-[a-z0-9]+' | sort -u | wc -l) (want 3)"
}
trial lib.rs        # -> MARK_lines=1 blame_states=3  CLEAN
trial src/lib.rs    # -> MARK_lines=2 blame_states=1  CORRUPT (verified)
```
- **Bug 1** (above): `src/lib.rs` -> duplicated `// MARK`, blame collapse.
- **Bug 2**: base file ending in `#[cfg(test)] mod tests { … }`, each thread adds a `pub fn` above the marker -> the merge nests one fn **inside** `mod tests` and duplicates the module + preamble; `rustc`: "unclosed delimiter". Same root cause.
- **Bug 3**: nested re-export manifest `pub mod prelude { pub use crate::a; }`, each thread adds `pub use crate::<x>;` inside it + a flat `pub fn` -> one re-export **escapes** `prelude` followed by a dangling `}`; "unexpected closing delimiter". (#477 fixed *flat top-level* `use`; nested-in-module re-exports merged alongside a sibling item still corrupt.) Same root cause.

## Shared-cause verdict: one bug class, two stacked defects
Bugs 1/2/3 share the same composition. **Highest-leverage fix = `reconstruct_merged_file` in `crates/semantic/src/merge_driver/reconstruct.rs`** — fixing the inter-item/postamble dedup makes the semantic fallback correct, neutralizing the corruption for all three even when the rebase over-conflicts. Separately, the rebase engine should not spuriously conflict on subdir paths (that's what selects the buggy fallback) — fixing only that hides 2/3 of the surface (root merges) but leaves the driver landmine for any legitimate 3-way.

## Close-the-class conformance harness (this whole class is invisible to the marker-based checks)
Add a property test in `crates/semantic/src/merge_driver/tests.rs` asserting, for every clean `semantic_three_way_merge`, on the **parsed** output:
1. **Item-set + nesting conservation:** merged top-level item set == union(base,ours,theirs) keys minus deletions, and each item's parent-module path equals the contributing side's (catches Bug 2/3).
2. **No invented/duplicated lines:** every output line outside a merged item body appears in exactly one side's inter-item ranges; `count(line in output) <= max_over_sides count(line)` for marker/comment lines (catches Bug 1).
3. **Output re-parses** (`ParsedFile::parse(output).is_some()`) — a clean merge producing unparseable output is by definition a bug.

Run across `{flat-top-level, trailing-mod, nested-pub-use} × {added-above, added-below, added-inside}`. Plus a CLI-level test: `thread refresh` on a subdirectory file must be byte-identical to the same content at repo root (pins the rebase half).

---
*Found while building a threads + semantic-merge + attribution demo. Verified on `main` @ `5332f61d`. Bug 4 (isolated-checkout attribution) and all 4 earlier cross-reference items do NOT reproduce on current main — superseded by #465/#478.*

### PR DESCRIPTION
**Pull request by thethorne48**
Closes #484. Supersedes the paused PR #487.

## Root-cause fix for the P0 silent merge-corruption
The original merge driver flattened the parse into a flat item list and re-derived container nesting/instance identity downstream — the source of the silent collapse of reopened/added same-name containers. PR #487's instance-ordinal patch leaked across 4 review rounds and its output-floor could be fooled, so we fix it at the root:

**Containers are now first-class tree nodes.** `impl`/`mod`/`trait`/`namespace` are emitted as `Item`s with explicit `children` edges (items.rs); the 3-way merge runs on the shallow tree via the existing `MatchKey` matcher — container identity falls out of the matcher, so there are **no ordinals to mis-assign**; reconstruction is a recursive pre-order emit. The instance-ordinal class of bugs becomes structurally impossible.

## Oracle
Ported the full conformance harness from the #487 branch (tests.rs +636): original heddle-merge-bugs.md P0 repros + separator matrix (back-to-back/comment/whitespace/item) + cross-side add-before/add-after/reorder + the round-4 prepend+append case — all pass by construction.

## Test plan
- [x] `cargo test -p heddle-semantic` (ported harness, all rounds) — agent-run
- [x] `cargo test --workspace` / `cargo clippy --locked -D warnings` — agent-run
- [ ] CI Check & test + Coverage green
- [ ] Codex review

---
<a href="https://app.blacksmith.sh/HeddleCo/codesmith/heddle/pr/506"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a> <a href="https://backend.blacksmith.sh/track/enable-autofix?expires=1783149640&installation_id=132405951&pr_number=506&repository=HeddleCo%2Fheddle&return_to=https%3A%2F%2Fgithub.com%2FHeddleCo%2Fheddle%2Fpull%2F506&signature=e3b8012e22eba4564770feff9cd755181d775713d485d3d73cabfb09c1c964fc"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/autofix-with-codesmith-light.svg">Autofix with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is disabled.</sup>

### CODE REVIEW

**Comment by thethorne48** [comment]
@codex please review — this is the P0 root-cause fix (closes #484, supersedes the paused #487). Containers are now first-class tree nodes (children edges); the 3-way merge runs on the parse tree via the existing MatchKey matcher, so container-instance identity falls out of the parse — no ordinals to mis-assign, the silent-collapse class is structurally impossible. Ported the full #487 conformance harness (original P0 repros + separator matrix + cross-side add/reorder + the round-4 prepend+append case) — all pass by construction. Check & test green.

File: crates/semantic/src/merge_driver/reconstruct.rs [hunk 1]
```diff
@@ -263,6 +281,103 @@ fn build_match_keys(seg: &FileSegments) -> Vec<MatchKey> {
...
+                let childset = child_key_set(it);
+                let used_set = used.entry(&it.key).or_default();
+                let mut best: Option<(usize, usize)> = None; // (overlap, base disc)
+                if let Some(cands) = base_by_key.get(&it.key) {
+                    for (disc, bitem) in cands {
+                        if used_set.contains(disc) {
+                            continue;
+                        }
+                        let overlap = childset.intersection(&child_key_set(bitem)).count();
+                        if overlap > 0 && best.is_none_or(|(o, _)| overlap > o) {
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/reconstruct.rs]: Fixed in 780dc95d: when no base candidate has child-overlap > 0 (the empty-container / fully-replaced-children case), the matcher now falls back to the **next unused base occurrence** of the same key (positional, by base source order) before minting a fresh discriminator — so empty same-name containers align to their base slot instead of being treated as new. Content-overlap stays the primary match; fresh discriminators only when no unused base slot remains. Extended the harness with the two-empty-blocks (one-sided edit / both-sided edit) + fully-replaced-children cases (Rust + C++), confirmed fail-pre/pass-post. Check & test green.

**Comment by thethorne48** [comment]
@codex please review the updates here — 780dc95d adds a positional/base-occurrence fallback for zero-overlap containers (empty or fully-replaced-children same-name blocks now align to their base slot instead of minting fresh discriminators), with harness cases for the two-empty-blocks one-sided/both-sided edits. Content-overlap matching stays primary. Check & test green.

File: crates/semantic/src/merge_driver/reconstruct.rs [hunk 2]
```diff
@@ -263,6 +281,157 @@ fn build_match_keys(seg: &FileSegments) -> Vec<MatchKey> {
...
+        // zero-overlap case, NOT a return to the per-everything ordinal model.
+        let mut out = Vec::with_capacity(side.len());
+        for (pos, it) in side.iter().enumerate() {
+            if container_keys.contains(&it.key) {
+                let disc = if let Some(d) = disc_of[pos] {
+                    d
+                } else {
+                    let used_set = used.entry(&it.key).or_default();
+                    let next_unused = base_by_key.get(&it.key).and_then(|cands| {
+                        cands.iter().map(|(d, _)| *d).find(|d| !used_set.contains(d))
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/reconstruct.rs:445]: Fixed in c2fc3784, two parts:
1. **Header-anchor alignment** — zero-overlap containers now align to their **true base occurrence** (by header/source position relative to base anchors), accounting for deleted/added earlier occurrences, instead of claiming the first unused base slot. The delete-earlier-edit-later case (ours deletes the first empty `impl Foo {}` + edits the second, theirs edits the separator) now keeps the survivor at its original location with both edits, separator not moved/duplicated.
2. **Input-grounded conservation floor** (`conserves_inputs`) — after a clean reconstruction it re-parses the output and checks item-set + nesting conservation **against the inputs** (not the merge's own resolved metadata); any residual mis-alignment that drops/invents/mis-nests an item or fails to re-parse routes to a **textual conflict** rather than silent corruption. So even an alignment edge case we haven't enumerated is now safe-by-conflict, never silent corruption. Harness extended (delete-earlier/prepend/reorder + floor cases, fail-pre/pass-post). Check & test green.

**Comment by thethorne48** [comment]
@codex please review the updates here — c2fc3784 header-anchors zero-overlap container alignment to the true base occurrence (handles delete-earlier/prepend/reorder), AND adds an input-grounded conservation floor (re-parse + item-set/nesting check against the inputs → textual conflict on any residual mis-alignment, never silent corruption). Harness extended with those cases + floor no-false-positive checks. Check & test green.

File: crates/semantic/src/merge_driver/reconstruct.rs [hunk 3]
```diff
@@ -440,6 +623,215 @@ impl EolPolicy {
...
+    sides: SideSources<'_>,
+    base: Option<&Item>,
+    ours: &Item,
+    theirs: &Item,
+    markers: ConflictMarkers<'_>,
+) -> (Option<Vec<u8>>, usize) {
+    let (header, hc) = merge3_text(
+        base.map(|b| header_bytes(b, sides.base)),
+        header_bytes(ours, sides.ours),
+        header_bytes(theirs, sides.theirs),
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/reconstruct.rs:806]: Fixed in 9fd2b467: the structural-merge entry is now guarded by a precondition check — `merge_container_3way` is only entered when ALL participating sides are containers WITH a body. A matched key that is a container on one side but a leaf on another (Python `class` + an added `decorated_definition` leaf; Rust `mod foo { ... }` → `mod foo;`) no longer enters the structural path (no more `header_bytes` unwrap-None panic) — it routes to a whole-item 3-way text merge, and to a normal conflict if that overlaps. This closes the whole 'structural-merge precondition violated' class. Tests added for the Python class+decorator and Rust mod{}→mod; cases (fail-pre panic / pass-post). Check & test green.

**Comment by thethorne48** [comment]
@codex please review the updates here — 9fd2b467 guards the structural-merge entry: a matched key that is a container on one side but a leaf on another (class+decorator, mod{}→mod;) no longer enters merge_container_3way (no panic) — it routes to whole-item text merge or conflict. Closes the structural-precondition-violated class; tests for both cases. Check & test green.

File: crates/semantic/src/merge_driver/reconstruct.rs [hunk 4]
```diff
@@ -440,6 +623,240 @@ impl EolPolicy {
...
+            return (theirs.to_vec(), 0);
+        }
+        if theirs == b {
+            return (ours.to_vec(), 0);
+        }
+    }
+    let b = base.unwrap_or(&[]);
+    materialize_segment(
+        text_hunk_merge_with_markers(b, ours, theirs, markers),
+        std::str::from_utf8(b).unwrap_or(""),
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/reconstruct.rs]: Fixed in 07ab098c: divergent add/add container headers/footers (base None, same key, differing bytes — e.g. ours `pub mod foo { ... }` vs theirs `mod foo { ... }`) now emit a **conflict** rather than going through `text_hunk_merge_with_markers` with an empty base (which diff3 treats as a clean concatenation). This mirrors the existing `resolve_item` divergent-add/add leaf handling. Identical add/add (both sides add the same thing) stays clean. I swept the sibling empty-base add/add paths in the merge_driver so none silently concatenates divergent insertions — the invariant now holds uniformly. Tests added for divergent + identical add/add containers (fail-pre concat / pass-post conflict). Check & test green.

**Comment by thethorne48** [comment]
@codex please review the updates here — 07ab098c makes divergent add/add container headers/footers emit a conflict instead of diff3's empty-base silent concat (mirrors resolve_item's leaf handling), and sweeps the sibling empty-base add/add paths so divergence is uniformly a conflict. Identical add/add stays clean. Tests for both. Check & test green.

File: crates/semantic/src/merge_driver/reconstruct.rs [hunk 5]
```diff
@@ -440,6 +623,253 @@ impl EolPolicy {
...
+    let tb = theirs.body.as_ref().expect("container");
+    // Empty base body uses bounds (0, 0): always valid for `sides.base`
+    // regardless of how short it is, and yields a single empty preamble so
+    // both sides' children weave in as additions.
+    let (base_children, base_bounds): (&[Item], (usize, usize)) = match base {
+        Some(b) => {
+            let bb = b.body.as_ref().expect("container");
+            (&bb.items, (bb.inner_start, bb.inner_end))
+        }
+        None => (&[], (0, 0)),
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/reconstruct.rs]: Fixed in 2bac0ae6 (close-the-class): a same-key container ADDED on both sides with no base anchor is no longer recursively body-merged (that's where the delimiter/header bugs lived — r4 divergent headers, r5 duplicated `{`). Instead the two added containers are compared as WHOLE units: identical full content → take one (clean); different → whole-container conflict. A recursive structural merge only happens when there's a real BASE container to anchor against. This subsumes r4's per-case logic and fixes r5 (`mod foo { fn a }` vs `mod foo { fn b }` → conflict, no module duplication). Tests added for disjoint-children / identical / divergent-header / base-anchored cases (fail-pre/pass-post). Check & test green.

**Comment by thethorne48** [comment]
@codex please review the updates here — 2bac0ae6 closes the r3/r4/r5 cluster structurally: a same-key container added on BOTH sides with no base anchor is compared as a whole unit (identical→take one, different→whole-container conflict) instead of recursively body-merging (which mis-wove headers/delimiters). Recursive structural merge is reserved for the base-anchored case. Tests for disjoint-children/identical/divergent-header/base-anchored. Check & test green.

File: crates/semantic/src/merge_driver/reconstruct.rs [hunk 6]
```diff
@@ -146,65 +182,57 @@ pub(crate) fn reconstruct_merged_file(
...
+            // slot — its surrounding content flows with its own items and its
+            // trailing content stays in the postamble. Bridging a lacking
+            // side to "the gap after its nearest prior item" used to pull the
+            // postamble (or a mid-sequence gap) in early, duplicating it —
+            // the heddle#484 Bug 1 (`// MARK` woven twice) / Bug 2 (module
+            // duplicated) class.
+            if let Some(&r) = side_idx_maps[s].get(key)
+                && emitted[s].insert(r)
+            {
                 segs[s] = Some(inter_slice(side_sources[s], &side_ranges[s], r));
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/reconstruct.rs:236]: Fixed in 2d611007: the structural-container reconstruction now emits the body preamble/delimiter exactly once — the base-anchored and add/add container paths funnel through a single delimiter-emit instead of each writing their own. And the input-grounded conservation floor (`conserves_inputs`) now also runs on the conflict branch, not just the clean branch, so a container-body divergence falls to a conflict rather than slipping through as a double- or zero-emit. CI green on 2d611007.

**Comment by thethorne48** [comment]
@codex please review the updates here — r6 addresses the empty-container-body-preamble finding (emit-once + floor-on-conflict), CI green on 2d611007.

File: crates/semantic/src/merge_driver/items.rs [hunk 7]
```diff
@@ -232,51 +426,76 @@ pub(crate) fn canonicalize_use_keys(
...
-    let poisoned = [&*base, &*ours, &*theirs].iter().any(|seg| {
-        seg.items
-            .iter()
-            .any(|item| matches!(item.use_identity, Some(UseIdentity::Unanalyzable)))
-    });
+    let mut poisoned = false;
+    for seg in [&*base, &*ours, &*theirs] {
+        visit_items(&seg.items, &mut |item| {
+            if matches!(item.use_identity, Some(UseIdentity::Unanalyzable)) {
+                poisoned = true;
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge_driver/items.rs]: Fixed in a13a0bf0: use-poisoning is now scope-keyed to the container-as-node region instead of a single file-wide flag — a nested unanalyzable glob (e.g. `mod m { use x::*; }`) poisons only its own scope, so disjoint top-level `use` additions no longer collapse into a spurious add/add conflict. Two regression tests: `rust_use_nested_glob_does_not_poison_disjoint_top_level_adds` (sibling regions merge clean) and `rust_use_nested_glob_still_poisons_its_own_scope` (own-scope conservatism preserved). CI green post-unblock.

**Comment by thethorne48** [comment]
@codex please review the updates here — r7 scope-keys the use-poison to the container region (close-the-class for the file-global poison), CI green.
